### PR TITLE
fix/ota-3865/Log connectivity restored after an interruption

### PR DESCRIPTION
**Pull request by Zee314159**
Draft code.

### CODE REVIEW

File: src/libaktualizr/primary/sotauptaneclient.cc [hunk 1]
```diff
@@ -948,15 +948,23 @@ bool SotaUptaneClient::putManifestSimple(const Json::Value &custom) {
...
   auto manifest = AssembleManifest();
   if (custom != Json::nullValue) {
     manifest["custom"] = custom;
   }
   auto signed_manifest = uptane_manifest.signManifest(manifest);
   HttpResponse response = http->put(config.uptane.director_server + "/manifest", signed_manifest);
   if (response.isOk()) {
+    if (!connect) {
+      LOG_INFO << "Connectivity is restored.";
+      sendEvent<event::LogConnectivityRestored>();
```
Comment (patrickvacek) [inline src/libaktualizr/primary/sotauptaneclient.cc]: I think just logging is enough. This is probably easier than we were thinking. Sending an event isn't a bad idea but for now we can skip it.

File: src/libaktualizr/uptane/uptane_test.cc [hunk 2]
```diff
@@ -282,6 +282,44 @@ TEST(Uptane, PutManifestError) {
...
+
+  http = std::make_shared<HttpFake>(temp_dir.Path());
+  auto events_channel = std::make_shared<event::Channel>();
+  std::function<void(std::shared_ptr<event::BaseEvent> event)> f_cb = process_events_LogConnectivityRestored;
+  events_channel->connect(f_cb);
+  sota_client = std_::make_unique<UptaneTestCommon::TestUptaneClient>(conf, storage, http, events_channel);
+  EXPECT_NO_THROW(sota_client->initialize());
+  EXPECT_TRUE(sota_client->putManifestSimple());
+  EXPECT_EQ(num_events_LogConnectivityRestored, 1);
+}
```
Comment (patrickvacek) [inline src/libaktualizr/uptane/uptane_test.cc]: This general test format looks fine, although capturing logging output is slightly challenging, and I'd suggest putting this in `uptane_network_test.cc`, although to be honest this isn't really Uptane-specific.

**Comment by Zee314159** [comment]
```

[2019-12-02 11:56:38.874602] [0x000000000404fcc0] [info]    Bootstrap empty SQL storage
4401
[2019-12-02 11:56:38.896849] [0x000000000404fcc0] [info]    Bootstraping DB to version 22
4402
[2019-12-02 11:56:39.130130] [0x000000000404fcc0] [trace]   meta with role root in repo director not present in db
4403
[2019-12-02 11:56:39.157554] [0x000000000404fcc0] [trace]   meta with role root in repo image not present in db
4404
[2019-12-02 11:56:39.214023] [0x000000000404fcc0] [trace]   device_id not present in db
4405
[2019-12-02 11:56:39.277134] [0x000000000404fcc0] [trace]   client_pkey not present in db
4406
[2019-12-02 11:56:39.316517] [0x000000000404fcc0] [trace]   client_cert not present in db
4407
[2019-12-02 11:56:39.347844] [0x000000000404fcc0] [trace]   ca_cert not present
4408
[2019-12-02 11:56:39.376881] [0x000000000404fcc0] [trace]   client_pkey not present in db
4409
[2019-12-02 11:56:39.837780] [0x000000000404fcc0] [trace]   post request body:{
4410
	"deviceId" : "tasty-raclette-116",
4411
	"ttl" : "logconnect"
4412
}
4413
[2019-12-02 11:56:39.858029] [0x000000000404fcc0] [trace]   response http code: 404
4414
[2019-12-02 11:56:39.858286] [0x000000000404fcc0] [trace]   response: 
4415
[2019-12-02 11:56:39.859687] [0x000000000404fcc0] [error]   Shared credential provisioning failed, response: 
4416
[2019-12-02 11:56:39.859988] [0x000000000404fcc0] [error]   Shared credential provisioning failed. Aborting initialization.
4417
../src/libaktualizr/uptane/uptane_network_test.cc:163: Failure
4418
Value of: result
4419
  Actual: false
4420
Expected: true
```

Any tips why my test failed at the first stage? I followed the `"noerrors"` test case to write this one. And `testing::internal::CaptureStdout` seems not like a right choice to capture LOG_INFO. Any other suggestions?

File: src/libaktualizr/uptane/uptane_network_test.cc [hunk 3]
```diff
@@ -139,6 +140,44 @@ TEST(UptaneNetwork, DownloadFailure) {
   EXPECT_TRUE(result.first);
 }
 
+/*
+ * Output a log when connectivity is restored.
+ */
+TEST(UptaneNetwork, LogConnectivityRestored) {
+  //  testing::internal::CaptureStdout();
+  RecordProperty("zephyr_key", "OTA-991,TST-158");
```
Comment (patrickvacek) [inline src/libaktualizr/uptane/uptane_network_test.cc]: We can leave out this `RecordProperty` thing for now. Maybe someday that'll be important again, but for now, it isn't.

File: src/libaktualizr/uptane/uptane_network_test.cc [hunk 4]
```diff
@@ -139,6 +140,44 @@ TEST(UptaneNetwork, DownloadFailure) {
...
+    KeyManager keys(store, conf.keymanagerConfig());
+    Initializer initializer(conf.provision, store, http, keys, {});
+    result = initializer.isSuccessful();
+  }
+  EXPECT_TRUE(result);
+  {
+    conf.provision.server = conf.provision.server.substr(conf.provision.server.size() - 2) + "11";
+    KeyManager keys(store, conf.keymanagerConfig());
+    Initializer initializer(conf.provision, store, http, keys, {});
+    result = initializer.isSuccessful();
```
Comment (patrickvacek) [inline src/libaktualizr/uptane/uptane_network_test.cc]: I like this idea of feigning a broken connection by changing the server URL. However, does it require re-initialization every time? I was hoping we could just test this with calls to `CheckUpdates`. At any rate, we almost certainly can avoid extra instantiations of `KeyManager`.

Comment (Zee314159) [reply src/libaktualizr/uptane/uptane_network_test.cc]: > At any rate, we almost certainly can avoid extra instantiations of KeyManager.

But `Initializer initializer(conf.provision, store, http, keys, {});` need `keys`. And from `doTestInit` it looks re-initialization is needed. I'm not sure though.

Comment (patrickvacek) [reply src/libaktualizr/uptane/uptane_network_test.cc]: I'm fairly confident that you can reuse the keys if not the entire initialization. To simulate a more realistic scenario, we should only initialize once and merely call `CheckUpdates` multiple times under different network conditions.

Comment (Zee314159) [reply src/libaktualizr/uptane/uptane_network_test.cc]: > we should only initialize once and merely call CheckUpdates multiple times under different network conditions.

How can I get under different network conditions if I don't use initialize? I didn't find the function I need.

File: src/libaktualizr/uptane/uptane_network_test.cc [hunk 5]
```diff
@@ -139,6 +140,37 @@ TEST(UptaneNetwork, DownloadFailure) {
...
+  conf.provision.primary_ecu_serial = "noerrors";
+  const std::string good_url = conf.provision.server;
+
+  auto http = std::make_shared<HttpClient>();
+  auto store = INvStorage::newStorage(conf.storage);
+
+  auto sota_client = std_::make_unique<SotaUptaneClient>(conf, store, http);
+  EXPECT_NO_THROW(sota_client->initialize());
+  EXPECT_NO_THROW(sota_client->fetchMeta());
+  EXPECT_NO_THROW(sota_client->checkUpdates());
```
Comment (patrickvacek) [inline src/libaktualizr/uptane/uptane_network_test.cc]: If you're using the SotaUptaneClient functions directly, you probably want to just call `initialize` once and then `fetchMeta` each time. `fetchMeta` is actually what `Aktualizr::CheckUpdates` calls, and `fetchMeta` calls `SotaUptaneClient::checkUpdates`. It also is what calls `putManifestSimple`, which is where your change is located, so just calling `SotaUptaneClient::checkUpdates` won't trigger it.

Comment (patrickvacek) [reply src/libaktualizr/uptane/uptane_network_test.cc]: BTW if you make this test case a `FRIEND_TEST` of `SotaUptaneClient` (see `sotauptaneclient.h` for other examples) you can directly manipulate the `Config` object stored in `SotaUptaneClient` and then you can bypass needing to re-instantiate or re-initialize `SotaUptaneClient`.

File: src/libaktualizr/uptane/uptane_network_test.cc [hunk 6]
```diff
@@ -139,6 +140,33 @@ TEST(UptaneNetwork, DownloadFailure) {
...
+  conf.storage.path = temp_dir.Path();
+  conf.provision.expiry_days = "noerrors";
+  conf.provision.primary_ecu_serial = "noerrors";
+  const std::string good_url = conf.provision.server;
+  auto http = std::make_shared<HttpClient>();
+  auto store = INvStorage::newStorage(conf.storage);
+
+  auto sota_client = std_::make_unique<SotaUptaneClient>(conf, store, http);
+  EXPECT_NO_THROW(sota_client->initialize());
+  result::UpdateCheck result = sota_client->fetchMeta();
```
Comment (Zee314159) [inline src/libaktualizr/uptane/uptane_network_test.cc]: Got this failure `HTTP/1.0 501 Unsupported method ('PUT')`. Is `auto http = std::make_shared<HttpClient>();` a susceptive target?

Comment (Zee314159) [reply src/libaktualizr/uptane/uptane_network_test.cc]: `HttpFake` seems to make no difference in local test.

**Comment by lbonn** [comment]
It looks like it's failing right now because the "hasupdates" HttpFake should be used with a secondary. 

Adding this line makes the fetchMeta call succeed. It would even be better to use `makeTestConfig` from `uptane_test_common.h` instead of this shared object between all tests.

```
UptaneTestCommon::addDefaultSecondary(conf, temp_dir, "secondary_ecu_serial", "secondary_hw");
```

**Comment by Zee314159** [comment]
> It looks like it's failing right now because the "hasupdates" HttpFake should be used with a secondary.
> Adding this line makes the fetchMeta call succeed. It would even be better to use makeTestConfig from uptane_test_common.h instead of this shared object between all tests.
> UptaneTestCommon::addDefaultSecondary(conf, temp_dir, "secondary_ecu_serial", "secondary_hw");

Thanks Laurent@lbonn ,yes, I tried your method, the fetchMeta call succeed, but when I try to simulate a broken connection, it seems still connected, and failed to directly manipulate the `Config object` stored in `SotaUptaneClient`. Any tips?

**Comment by lbonn** [comment]
@Zee314159 Modifying the config object might work, or not if it has been copied before by an internal component.

The most straightforward thing that comes to my mind is to subclass HttpFake and add a new method to toggle a switch that causes all the requests to fail.

File: src/libaktualizr/uptane/uptane_network_test.cc [hunk 7]
```diff
@@ -139,6 +141,85 @@ TEST(UptaneNetwork, DownloadFailure) {
...
+      return HttpResponse(url, 200, CURLE_OK, "");
+    }
+  }
+
+  HttpResponse post(const std::string &url, const Json::Value &data) override {
+    if (!connectSwitch) {
+      (void)data;
+      return HttpResponse(url, 503, CURLE_OK, "");
+    } else {
+      if (url.find("/devices") != std::string::npos || url.find("/director/ecus") != std::string::npos || url.empty()) {
```
Comment (lbonn) [inline src/libaktualizr/uptane/uptane_network_test.cc]: Is there a reason why we can't call `HttpFake::post()` here? Copying this logic here, would complicate maintenance.

File: src/libaktualizr/uptane/uptane_network_test.cc [hunk 8]
```diff
@@ -139,6 +141,85 @@ TEST(UptaneNetwork, DownloadFailure) {
...
+      return HttpFake::get(url, maxsize);
+    }
+  }
+
+  HttpResponse put(const std::string &url, const Json::Value &data) override {
+    if (!connectSwitch) {
+      (void)data;
+      return HttpResponse(url, 504, CURLE_OK, "");
+    } else {
+      last_manifest = data;
```
Comment (lbonn) [inline src/libaktualizr/uptane/uptane_network_test.cc]: Can we also call `HttpFake::put()`?

**Comment by lbonn** [comment]
Looking good apart from the two small comments. Also, before we merge, could you edit the commit message so that it describes the changes better?

File: src/libaktualizr/uptane/uptane_network_test.cc (hunk 7 above)
Comment (Zee314159) [reply src/libaktualizr/uptane/uptane_network_test.cc]: I'm not sure, but if will fail if I use the original. I'll have a test.

Comment (Zee314159) [reply src/libaktualizr/uptane/uptane_network_test.cc]:
```
file served: "/tmp/aktualizr-4d8c-c7cb-fcae-9383/619b-47da-dir/repo/targets_hasupdates.json"
4429
[2019-12-09 12:59:48.635639] [0x000000000404fcc0] [info]    2 new updates found in Uptane metadata.
4430
[2019-12-09 12:59:48.659888] [0x000000000404fcc0] [info]    got UpdateCheckComplete event
4431
[2019-12-09 12:59:50.232787] [0x000000000404fcc0] [error]   Failed to update director metadata: 
4432
[2019-12-09 12:59:50.236768] [0x000000000404fcc0] [info]    got UpdateCheckComplete event
4433
../src/libaktualizr/uptane/uptane_network_test.cc:217: Failure
4434
Expected: (std::string::npos) != (testing::internal::GetCapturedStdout().find("Connectivity is restored.")), actual: 18446744073709551615 vs 1844674407370955161
```

Comment (lbonn) [reply src/libaktualizr/uptane/uptane_network_test.cc]: Well, you still need to handle the failure case. In the last version you've pushed, `put` and `post` are the same as in HttpFake.

If you follow the same pattern that what you did for `get()`, it should work.

e.g:

```
HttpResponse post(const std::string &url, const Json::Value &data) override {
 if (!connectSwitch) {
   (void)data;
   return HttpResponse(url, 503, CURLE_OK, "");
 } else {
   return HttpFake::post(url, data);
  }
}
```

Comment (Zee314159) [reply src/libaktualizr/uptane/uptane_network_test.cc]: Oh, I see, thank you!

**Comment by lbonn** [comment]
It merged in master but the PR didn't get closed somehow :open_hands:

**Pull request closed without merging**